File: retroravelry/src/main/java/com/caseykulm/retroravelry/models/pattern/FullPattern.java
package com.caseykulm.retroravelry.models.pattern;

import com.caseykulm.retroravelry.models.craft.ListCraft;
import com.caseykulm.retroravelry.models.pack.FullPack;
import com.caseykulm.retroravelry.models.patternattribute.PatternAttribute;
import com.caseykulm.retroravelry.models.patternauthor.PatternAuthor;
import com.caseykulm.retroravelry.models.patterncategory.PatternCategory;
import com.caseykulm.retroravelry.models.photo.Photo;
import com.caseykulm.retroravelry.models.printing.ListPrinting;
import com.caseykulm.retroravelry.models.yarnweight.ListYarnWeight;
import com.google.auto.value.AutoValue;
import java.util.Date;
import java.util.List;
import java.util.Map;

@AutoValue public abstract class FullPattern {

  public abstract int comments_count();
  public abstract ListCraft craft();
  public abstract String currency();
  public abstract String currency_symbol();
  public abstract float difficulty_average();
  public abstract int difficulty_count();
  public abstract boolean downloadable();
  public abstract int favorites_count();
  public abstract boolean free();
  public abstract String gauge();
  public abstract String gauge_description();
  public abstract String gauge_divisor();
  public abstract String gauge_pattern();
  public abstract int id();
  public abstract String name();
  public abstract String notes_html();
  public abstract String notes();
  public abstract FullPack packs();
  public abstract List<PatternAttribute> pattern_attributes();
  public abstract PatternAuthor pattern_author();
  public abstract List<PatternCategory> pattern_categories();
  public abstract String pattern_needle_sizes();
  public abstract boolean pdf_in_library();
  public abstract String pdf_url();
  public abstract String permalink();
  public abstract Map<String, String> personal_attributes();
  public abstract List<Photo> photos();
  public abstract String price();
  public abstract List<ListPrinting> printings();
  public abstract int product_id();
  public abstract int projects_count();
  public abstract Date published();
  public abstract int queued_projects_count();
  public abstract float rating_average();
  public abstract int rating_count();
  public abstract boolean ravelry_download();
  public abstract String row_gauge();
  public abstract String sizes_available();
  public abstract String url();
  public abstract List<Integer> volumes_in_library();
  public abstract int yardage();
  public abstract String yardage_description();
  public abstract int yardage_max();
  public abstract ListYarnWeight yarn_weight();
  public abstract String yarn_weight_description();

}

File: retroravelry/src/main/java/com/caseykulm/retroravelry/models/pattern/FullPattern.kt
package com.caseykulm.retroravelry.models.pattern

import com.caseykulm.retroravelry.models.craft.ListCraft
import com.caseykulm.retroravelry.models.needles.PatternNeedleSize
import com.caseykulm.retroravelry.models.pack.FullPack
import com.caseykulm.retroravelry.models.patternattribute.PatternAttribute
import com.caseykulm.retroravelry.models.patternauthor.PatternAuthor
import com.caseykulm.retroravelry.models.patterncategory.PatternCategory
import com.caseykulm.retroravelry.models.personalattribute.PersonalAttribute
import com.caseykulm.retroravelry.models.photo.Photo
import com.caseykulm.retroravelry.models.printing.ListPrinting
import com.caseykulm.retroravelry.models.yarnweight.ListYarnWeight
import java.util.Date

data class FullPattern(
    val comments_count: Int,
    val craft: ListCraft?,
    val currency: String?,
    val currency_symbol: String?,
    val difficulty_average: Float,
    val difficulty_count: Int,
    val downloadable: Boolean,
    val favorites_count: Int,
    val free: Boolean,
    val gauge: String?,
    val gauge_description: String?,
    val gauge_divisor: String?,
    val gauge_pattern: String?,
    val id: Int,
    val name: String?,
    val notes_html: String?,
    val notes: String?,
    val packs: List<FullPack>?,
    val pattern_attributes: List<PatternAttribute>?,
    val pattern_author: PatternAuthor?,
    val pattern_categories: List<PatternCategory>?,
    val pattern_needle_sizes: List<PatternNeedleSize>?,
    val pdf_in_library: Boolean,
    val pdf_url: String?,
    val permalink: String?,
    val personal_attributes: PersonalAttribute?,
    val photos: List<Photo>?,
    val price: String?,
    val printings: List<ListPrinting>?,
    val product_id: Int,
    val projects_count: Int,
    //TODO Parse this into a recognized date format, currently returning YYYY/MM/DD instead of standard YYYY-MM-DD
    val published: String?,
    val queued_projects_count: Int,
    val rating_average: Float,
    val rating_count: Int,
    val ravelry_download: Boolean,
    val row_gauge: String?,
    val sizes_available: String?,
    val url: String?,
    val volumes_in_library: List<Int>?,
    val yardage: Int,
    val yardage_description: String?,
    val yardage_max: Int,
    val yarn_weight: ListYarnWeight?,
    val yarn_weight_description: String?
)
